src/Evolution/DiscontinuousGalerkin/Actions/ApplyBoundaryCorrections.hpp

```diff
@@ -231,46 +394,74 @@ bool receive_boundary_data_local_time_stepping(
+                                              element);
+
+          return boundary_data_ptr.get();
+        },
+        box,
+        make_not_null(
+            &tuples::get<
+                evolution::dg::Tags::BoundaryCorrectionAndGhostCellsInbox<Dim>>(
+                *inboxes)),
+        db::get<domain::Tags::Element<Dim>>(*box));
+  } else {
+    inbox = &tuples::get<
+        evolution::dg::Tags::BoundaryCorrectionAndGhostCellsInbox<Dim>>(
+        *inboxes);
+  }
+
   const bool have_all_intermediate_messages = db::mutate<
       evolution::dg::Tags::MortarDataHistory<Dim,
                                              typename dt_variables_tag::type>,
       evolution::dg::Tags::MortarNextTemporalId<Dim>,
       domain::Tags::NeighborMesh<Dim>>(
       [&inbox, &needed_time](
-          const gsl::not_null<
-              std::unordered_map<Key,
-                                 TimeSteppers::BoundaryHistory<
-                                     evolution::dg::MortarData<Dim>,
-                                     evolution::dg::MortarData<Dim>,
-                                     typename dt_variables_tag::type>,
-                                 boost::hash<Key>>*>
+          const gsl::not_null<std::unordered_map<
+              Key,
+              TimeSteppers::BoundaryHistory<evolution::dg::MortarData<Dim>,
+                                            evolution::dg::MortarData<Dim>,
+                                            typename dt_variables_tag::type>,
+              boost::hash<Key>>*>
               boundary_data_history,
           const gsl::not_null<
               std::unordered_map<Key, TimeStepId, boost::hash<Key>>*>
-              mortar_next_time_step_ids,
-          const gsl::not_null<DirectionalIdMap<Dim, Mesh<Dim>>*>
-              neighbor_mesh,
+              mortar_next_time_step_id,
+          const gsl::not_null<DirectionalIdMap<Dim, Mesh<Dim>>*> neighbor_mesh,
           const Element<Dim>& element) {
         // Remove neighbor meshes for neighbors that don't exist anymore
         domain::remove_nonexistent_neighbors(neighbor_mesh, element);
 
         // Move received boundary data into boundary history.
-        for (auto& [mortar_id, mortar_next_time_step_id] :
-             *mortar_next_time_step_ids) {
-          if (mortar_id.id() == ElementId<Dim>::external_boundary_id()) {
-            continue;
-          }
-          while (needed_time(mortar_next_time_step_id)) {
-            const auto time_entry = inbox.find(mortar_next_time_step_id);
-            if (time_entry == inbox.end()) {
-              return false;
-            }
-            const auto received_mortar_data =
-                time_entry->second.find(mortar_id);
-            if (received_mortar_data == time_entry->second.end()) {
-              return false;
-            }
-
+        ASSERT(inbox != nullptr, "The inbox pointer should not be null.");
+        for (auto received_data = inbox->begin();
```

Comment:
The rest of this function looks like it's based off a version of the code from before #5864.  It needs to incorporate those changes.  I think the intended changes here are mostly changing `inbox` to a pointer, so that shouldn't be too difficult.